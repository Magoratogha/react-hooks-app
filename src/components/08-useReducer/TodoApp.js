import { Fragment, useCallback, useEffect, useReducer } from "react";
import todoReducer from "./todoReducer";
import TodoList from "./TodoList";
import TodoAdd from "./TodoAdd";
import './styles.css';

const init = () => {
    return JSON.parse(localStorage.getItem('todos')) || [];
};

function TodoApp() {

    const [todos, dispatch] = useReducer(todoReducer, [], init);

    const handleAddTodo = useCallback((newTodo) => {
        dispatch({
            type: 'add',
            payload: newTodo
        });
    }, []);

    const handleDelete = useCallback((todoId) => {
        dispatch({
            type: 'delete',
            payload: todoId
        });
    }, []);

    const handleToggle = useCallback((todoId) => {
        dispatch({
            type: 'toggle',
            payload: todoId
        });
    }, []);

    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos));
    }, [todos]);

    return (
        <Fragment>
            <h1>TodoApp ({todos.length})</h1>
            <hr></hr>
            <div className="row">
                <div className="col-7">
                    <TodoList
                        todos={todos}
                        handleDelete={handleDelete}
                        handleToggle={handleToggle} />
                </div>
                <div className="col-5">
                    <TodoAdd
                        handleAddTodo={handleAddTodo} />
                </div>
            </div>
        </Fragment>
    );
}

export default TodoApp;