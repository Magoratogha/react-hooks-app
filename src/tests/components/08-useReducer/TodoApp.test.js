import { render } from "@testing-library/react";
import TodoApp from "../../../components/08-useReducer/TodoApp";

describe('Test on <TodoApp />', () => { 

    test('should render properly', () => {
        const { container } = render(<TodoApp />);
        expect(container).toMatchSnapshot();
    });
});