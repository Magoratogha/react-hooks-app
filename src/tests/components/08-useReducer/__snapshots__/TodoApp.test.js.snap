// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on <TodoApp /> should render properly 1`] = `
<div>
  <h1>
    TodoApp (
    0
    )
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-7"
    >
      <ul
        class="list-group list-group-flush"
      />
    </div>
    <div
      class="col-5"
    >
      <h4>
        Add todo
      </h4>
      <hr />
      <form>
        <input
          autocomplete="off"
          class="form-control"
          name="description"
          placeholder="To do..."
          type="text"
          value=""
        />
        <div
          class="d-grid mt-2"
        >
          <button
            class="btn btn-outline-dark btn-sm"
            disabled=""
            type="submit"
          >
            Add
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
